ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/attestation/MatchingDataAttestationGroupV2.java

```diff
@@ -0,0 +1,520 @@
+      final long timeLimitNanos,
+      final Supplier<Stream<PooledAttestation>> candidatesStreamSupplier) {
+    return Spliterators.spliteratorUnknownSize(
+        createAggregatingIterator(timeLimitNanos, candidatesStreamSupplier), 0);
+  }
+
+  /**
+   * Returns true if there are no attestations in this group.
+   *
+   * @return true if this group is empty.
+   */
+  public boolean isEmpty() {
+    return attestationsByValidatorCount.isEmpty() && singleAttestationsByCommitteeIndex.isEmpty();
+  }
+
+  public int size() {
+    return attestationsByValidatorCount.values().stream().mapToInt(Set::size).sum()
+        + singleAttestationsByCommitteeIndex.values().stream().mapToInt(Set::size).sum();
+  }
+
+  /**
+   * Updates includedValidators bits and removes any attestation from this group whose aggregation
+   * bits have all been seen. Needs a write lock as it modifies state.
+   *
+   * @param attestation the attestation to logically remove from the pool.
+   */
+  public int onAttestationIncludedInBlock(final UInt64 slot, final Attestation attestation) {
+    int numRemoved;
+    writeLock.lock();
+    try {
+      // Record validators in attestation as seen in this slot
+      includedValidatorsBySlot.compute(
+          slot,
+          (__, includedValidators) -> {
+            if (includedValidators == null) {
+              return AttestationBits.of(attestation, committeesSize);
+            }
+            // Mutate includedValidators aggregator under write lock
+            includedValidators.or(attestation);
+            return includedValidators;
+          });
+
+      // Check if already seen before modifying the main includedValidators
+      if (includedValidators.isSuperSetOf(attestation)) {
+        // We've already seen and filtered out all of these bits, nothing to do
+        return 0;
+      }
+      // Mutate main includedValidators under write lock
+      includedValidators.or(attestation);
+
+      // Calculate size *before* removal for accurate delta.
+      int sizeBefore = size();
+
+      // Remove fully covered attestations using removeIf on the concurrent sets
+      attestationsByValidatorCount
+          .values()
+          .forEach(
+              candidates ->
+                  candidates.removeIf(
+                      candidate -> includedValidators.isSuperSetOf(candidate.bits())));
+
+      // Also remove empty sets from the outer map for cleanup
+      attestationsByValidatorCount.entrySet().removeIf(entry -> entry.getValue().isEmpty());
+
+      singleAttestationsByCommitteeIndex.forEach(
+          (committeeIndex, singleAttestations) -> {
+            singleAttestations.removeIf(
+                singleAttestation -> includedValidators.isSuperSetOf(singleAttestation.bits()));
+            if (singleAttestations.isEmpty()) {
+              singleAttestationsByCommitteeIndex.remove(committeeIndex);
+            }
+          });
+
+      int sizeAfter = size();
```

Comment:
could either call final here or just use size in computation of `numRemoved`

Reply:
yes, can be simplified!